### PR TITLE
Documentation

### PR DESCRIPTION
**Pull request by avaidyam**
Added some documentation to layout constraints and fixed a bug or two. Please do let me know if something in the docs is not clear or needs elaboration.

### CODE REVIEW

File: lib/UIKit/TUILayoutConstraint.m [hunk 1]
```diff
@@ -74,26 +83,55 @@ + (id)constraintWithAttribute:(TUILayoutConstraintAttribute)attr
...
 }
 
-- (CGFloat)transformValue:(CGFloat)original {
-	id transformed = [self.valueTransformer transformedValue:[NSNumber numberWithFloat:original]];
-	return [transformed floatValue];
+- (CGFloat)transformValue:(CGFloat)source {
+    if(self.valueTransformer) {
+        id transformed = [self.valueTransformer transformedValue:[NSNumber numberWithFloat:source]];
+        return [transformed floatValue];
+    } else return (source * self.scale) + self.offset;
```
Comment (jspahrsummers) [inline lib/UIKit/TUILayoutConstraint.m]: Do you mind putting this `return` on its own line (for consistency with the `if` above)?

**Comment by jspahrsummers** [comment]
Awesome! Just the one note, and then this will be good to go.

File: lib/UIKit/TUILayoutConstraint.m (hunk 1 above)
Comment (avaidyam) [inline lib/UIKit/TUILayoutConstraint.m]: Graahhh! I always do this =___= I'm sorry, it's been re-commited.

**Comment by avaidyam** [comment]
Yep, fixed now. :D

**Comment by jspahrsummers** [comment]
Thanks!

**Comment by jwilling** [comment]
```
+ (id)constraintWithAttribute:(TUILayoutConstraintAttribute)attr
                   relativeTo:(NSString *)source
                    attribute:(TUILayoutConstraintAttribute)srcAttr;
```

...always hits the assertion that there is an invalid scale and offset, which happens to be odd because supposedly it's evaluating that `0.0` is greater than `0.0f`.

**Comment by avaidyam** [comment]
Wait, haha, my bad XD Another commit incoming!

**Comment by jwilling** [comment]
Are you really testing this stuff? ;)

**Comment by avaidyam** [comment]
I'm using block transformers in my app xD A lot cooler, if you ask me. But yeah, didn't test this last commit :P (forgive me!)

## 

Aditya.

**Comment by jwilling** [comment]
Are you going to put an example of a block transformer in the documentation as well? I don't see the purpose in using a transformer in regards to layout constraints for some reason.

**Comment by avaidyam** [comment]
Ah, smart idea, will add that now before committing :D

## 

Aditya.

**Comment by jwilling** [comment]
And also, it seems that constraints bound to the superview aren't automatically adjusted when the superview resizes. Isn't that one of the most fundamental aspects of this whole thing?

**Comment by avaidyam** [comment]
Check in my fork for a commit; will add in a few more new classes and then pull request. Would you mind providing a sample where this happens? I can fix it up faster then. :]

**Comment by jwilling** [comment]
```
TUILayoutConstraint *pin = [TUILayoutConstraint constraintWithAttribute:TUILayoutConstraintAttributeMaxXMaxY relativeTo:@"superview" attribute:TUILayoutConstraintAttributeMaxXMaxY];
[splitView addLayoutConstraint:pin];
```

That works appropriately when the view first loads, however when the superview is resized the constraint does not update as expected.

**Comment by avaidyam** [comment]
Ahhh, figured out the issue, thanks! I'll push a commit soon :]

**Comment by jwilling** [comment]
Sweet.

**Comment by avaidyam** [comment]
A temporary fix: add a layout name to the superview and use that as a constraint rather than @"superview". Something about the superview not being registered as a valid layout view xD

And if you wanted the view to fill the superview perfectly, MaxXMaxY only aligns the vertices... you need Frame.

**Comment by jspahrsummers** [comment]
@galaxas0 Maybe you could create some unit tests or a sample project, to simultaneously test the API and demonstrate how to use it. Something like that would help catch bugs.

**Comment by avaidyam** [comment]
That's smart. I should think smarter. ;_____; Unfortunately, i'll have to rush it out by tomorrow then, as I'll be on a trip for a good 3 weeks… Please let me know what you'd want me to include, like certain examples, etc. :D

## 

Aditya.

**Comment by jspahrsummers** [comment]
@galaxas0 Seems like the biggest thing right now is just making sure that there's test or example code covering the different types of constraints.

I'd say verifying the different initializers is the most important, followed by verifying different attribute groups (side, corner, frame), followed by verifying each individual attribute within those groups. However much you can get done in that order would be awesome. :)

**Comment by avaidyam** [comment]
Alright, then I'll commit a few non-constraint related things out into my fork right now, to get those out of the way. By the way, what time is it for you there? It's 12:30 am for me xD

## 

Aditya.

**Comment by jwilling** [comment]
@galaxas0 Were you ever going to fix the superview bug? That's pretty serious.

**Comment by avaidyam** [comment]
Yeah, sorry, i'm pretty tight on time. I did a blitz to get a few changes in my "fork" into the main TwUI fork. I'll get to it in 3 days or so (traveling across the globe with no Macbook :[ )

**Comment by jwilling** [comment]
Ah man, I was hoping this would make my life easier making my split view but it doesn't really help, sadly. It's just not up to snuff with AppKit. Main two problems are the fact that you can't have flexible constraints (such as width>=100), and that you can't write with visual formatting. I know the latter would be quite hard to replicate and is probably not worth the time.

**Comment by avaidyam** [comment]
Well I can't do much as a one man army, can I? I'm at the airport right now too, so you may have to wait a bit for the superview glitch, the flexible constraints (this isn't too hard the way I've constructed the system). I've been cracking away at the ASCII art, and so far, I've replicated the `NSDictionaryOfVariableBindings` macro, and I've been thinking about how to remove the layout name mess, so you can just pass a view straight to it. Coming soon, but I'll obviously need help :]

**Comment by jwilling** [comment]
Hey, I wasn't attacking the work you've done so far. Admirable for working by yourself. Just saying what would need to be done at least to be comparable with AppKit. Keep it up. ;)

**Comment by avaidyam** [comment]
Haha, no worries, and thank you! I was just making a plea to the community is all :P

**Pull request merged**